### PR TITLE
Instance store check for encryption removal

### PR DESCRIPTION
**Pull request by blyles**
This is to fix hopefully the use_store_instance volumes for ephemeral, looking through the code it doesn't seem to take the encryption into account with use_raw_storage as it creates the disk.

Type, size and iops aren't applicable but the code checks the encrypted setting globally for ebs volumes which doesn't go through the system.  From what I can tell, removing ` || encrypted` from the code has no material impact on the tree under use_instance_storage for ephemeral for disk generation.

The spec's also dont' test for this , they only test for the existence of size leading me further to believe this isn't a worth while gate since it does a global settings check.  Would anyone be able to comment on this flag in the case of instance_storage?

 ```
def initialize(ephemeral_disk, global_config)
        super(ephemeral_disk)

        @encrypted = global_config.aws.encrypted  <--- It's pulling from here.
        @kms_key_arn = global_config.aws.kms_key_arn

        if ephemeral_disk
          @use_instance_storage = !!ephemeral_disk['use_instance_storage'] || false

          @use_root_disk = !!ephemeral_disk['use_root_disk'] if ephemeral_disk.key?('use_root_disk')

          @encrypted = !!ephemeral_disk['encrypted'] if ephemeral_disk.key?('encrypted')
          @kms_key_arn = ephemeral_disk['kms_key_arn'] if ephemeral_disk.key?('kms_key_arn')
        end
      end
```

And removing encrypted makes sense rather then interrupting the entire flow of ebs creation.

### CODE REVIEW

**Comment by jpalermo** [comment]
Hi @blyles, sorry for the delay. I don't have any experience with encrypting disks so it took me a while to dig into this.

From what I can tell (and I very well might be wrong), the `|| encrypted` is currently intended to throw an error because if a user specifies cloud properties of:
```
ephemeral_disk:
  use_instance_storage: true
  encrypted: true
```
the `encrypted: true` won't do anything because the only functionality in the CPI for encrypting a disk is at disk creation time, and if you are using the instance storage, we don't get a chance to actually encrypt that disk when the VM is created.

So the error that's being thrown is to prevent the user from thinking the disk is encrypted when it is not.

I feel like I'm missing a piece of the puzzle here though (much as I was for the last PR), can you give more details for the setup you're trying to create?

Are you saying the global `encrypted` property bleeds into this `ephemeral_disk.encrypted` property?

**Comment by blyles** [comment]
@jpalermo yup that's exactly what's going on I feel, that the global encrypted setting is blocking using storage optimized (local nvme from being setup).  We managed to get unit tests going and found that the encrypted doesn't seem to have any bearing when you set it globally on nvme attached (storage optimized) and ultimately want to do the following.

use i3en, i4i, m6id, c6id, r6id with ephemeral on instance storage and not using ebs, we have several deployments that are latency sensitive and the feature use_instance_storage seems the best place to go.

This line seems counter intuitive.

https://github.com/cloudfoundry/bosh-aws-cpi-release/blob/23e4e3ede33e1fd27369abed897f1cefa127a946/src/bosh_aws_cpi/lib/cloud/aws/cloud_props.rb#L185

Especially seeing that it seems that most current gens are encrypted at rest by default.  https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html.
 
I'm currently working through full e2e to get more answers on what's the go with InstanceStorage encrypted property, but at first glance the unit tests/spec tests don't even look for encrypted as it stands right now for use_instance_storage.

https://github.com/cloudfoundry/bosh-aws-cpi-release/blob/23e4e3ede33e1fd27369abed897f1cefa127a946/src/bosh_aws_cpi/lib/cloud/aws/block_device_manager.rb#L128 Exits the code leaving encrypted to default/empty for the disk mapping.

At the end of the day, var/vcap data on the nvme instead of ebs.  Hopefully this is actually logical and purposeful.

**Comment by blyles** [comment]
https://github.com/cloudfoundry/bosh-aws-cpi-release/blob/23e4e3ede33e1fd27369abed897f1cefa127a946/src/bosh_aws_cpi/lib/cloud/aws/block_device_manager.rb#L87

I think there may be an issue here too for when attempting to use, use_instance_storage, as a lot of the current gens have nvme requirements for ebs and aren't ebs only.

I thought this may have been as simple as the encrypted check was being over used, but a combination of the NVME families and expected output for ebs only crossed with vms that support instance storage may lead to unpredictable results.

**Comment by jpalermo** [comment]
It looks like `use_instance_storage` is only pulled from the `ephemeral_disk` cloud properties, so I'm guessing you're setting that somewhere (maybe on the vm_type in cloud_config?).

It seems like you can override the global `encrypted` property by reconfiguring it on the `ephemeral_disk` config: https://github.com/cloudfoundry/bosh-aws-cpi-release/blob/abcacf638cbc8bb1a6b2f8e5ba63a35a85ef054f/src/bosh_aws_cpi/lib/cloud/aws/cloud_props.rb#L193

Have you tried that yet?

**Comment by blyles** [comment]
Yes, setting it in the vm type, the full definition looks like this

```
- cloud_properties:
    ephemeral_disk:
      use_instance_storage: true #beta
    instance_type: i3en.large
  name: i3en.large-instance
```

Attempting to attach encrypted: true or encrypted at the vm type level as false, fails with time out and caused us to reach out on slack / cf, we're going to give it another whirl just incase we missed something.  But further testing says that any NVME family machine that has EBS only = false.  Needs more correction anyway.

Instance type | EBS only | NVMe EBS | Instance store | Placement group | Enhanced networking
-- | -- | -- | -- | -- | --
I3en | No | Yes | NVMe * | Yes | ENA
I3 | No | No | NVMe * | Yes | ENA

I have a few hours to spare this evening going to hopefully build out the remainder of the tests, the local rspec doesn't set encrypted globally.

**Comment by blyles** [comment]
Removed the encrypted check, seems our bosh agent has other problems - The only suggestion I changed was fixing the nvme1n1 like was pointed out.

**Comment by blyles** [comment]
We've just run more tests with;

```
- cloud_properties:
    ephemeral_disk:
      use_instance_storage: true
      encrypted: false
    instance_type: i3en.xlarge
  name: i3en.xlarge
  
```
And it creates the VM we can see it with /dev/xvda mounted and assigned but its looking to time out again, We still suspect its attempting to disable encryption causing issues with the global config setting to true.  Its got IP, everything in aws console....but...

- DirectorJobRunner: Failed to create/contact VM : #<Bosh::Director::RpcTimeout: Timed out pinging VM 'i-074e8853684ec7416' with agent 'e9fd2ce2-d9da-4212-aa3f-062714e29cfc' after 600 seconds>

We just get this over and over every time we try.  

@encrypted = !!ephemeral_disk['encrypted'] if ephemeral_disk.key?('encrypted') I feel like it gets bent out of shape in here, with the global config set to encrypted and trying to disable it makes it go funny along the tree - just can't see it.  I'm hoping i can get my code in e2e integration working, unit tests and all that are working but we've been yet unable to get e2e fired up properly which is hindering us.

Regards to logs, I'm having to enlist someone else with a bit more experience then me in bosh dissection.

**Comment by blyles** [comment]
At this point, here's where we are at with this PR we may close it until we can discover the origin of the failure with encryption in play.  And open a separate PR fixing the nvme1n1 mapping issue we created for when encryption is turned off.

Example configs.

Success while encryption is off by default (fresh director) and check in place

```
- cloud_properties:
    ephemeral_disk:
      use_instance_storage: true
    instance_type: i3en.xlarge
  name: i3en.xlarge
```
  
And fails with if check is in place and encryption is on globally.

```
- cloud_properties:
    ephemeral_disk:
      use_instance_storage: true
      **encrypted: false**
    instance_type: i3en.xlarge
  name: i3en.xlarge
```

Check **in place** for invalid use storage config.
If global encryption is on, both fail
If global config encryption is off both succeed.

Check **not in place for** invalid use storage config.
Pending further testing.

File: src/bosh_aws_cpi/lib/cloud/aws/block_device_manager.rb [hunk 1]
```diff
@@ -155,8 +155,10 @@ def first_raw_ephemeral_device
       case @virtualization_type
 
         when 'hvm'
-          if instance_type =~ /^i3\./ || instance_type =~ /^i3en\./
+          if instance_type =~ /^i3\./
             '/dev/nvme0n1'
+          elsif instance_type =~ /^i3en\./
```
[the remark this answers is not in the document]
Comment (blyles) [reply src/bosh_aws_cpi/lib/cloud/aws/block_device_manager.rb:160]: Ill be putting this in a separate PR to fix i think - this check here should be in place for all NVME Storage optimized instances with local NVME attached volumes.

**Comment by blyles** [comment]
We're going to rebuild this all out for a few reasons, going to close and try something else as an approach apologies been on leave.

**Pull request closed without merging**